mkosi/vmspawn.py

```diff
@@ -0,0 +1,123 @@
+import contextlib
+import os
+import sys
+from pathlib import Path
+
+from mkosi.config import (
+    MkosiArgs,
+    MkosiConfig,
+    OutputFormat,
+    QemuFirmware,
+    yes_no,
+)
+from mkosi.log import die
+from mkosi.partition import finalize_root, find_partitions
+from mkosi.qemu import (
+    KernelType,
+    apply_runtime_size,
+    copy_ephemeral,
+    finalize_qemu_firmware,
+)
+from mkosi.run import (
+    run,
+)
+from mkosi.types import PathString
+
+
+def run_vmspawn(args: MkosiArgs, config: MkosiConfig) -> None:
+    if config.output_format not in (OutputFormat.disk, OutputFormat.directory):
+        die(f"{config.output_format} images cannot be booted in systemd-vmspawn")
+
+    if config.qemu_firmware == QemuFirmware.bios:
+        die("systemd-vmspawn cannot boot BIOS firmware images")
+
+    if config.qemu_cdrom:
+        die("systemd-vmspawn does not support CD-ROM images")
+
+    kernel = config.qemu_kernel
+
+    if kernel and not kernel.exists():
+        die(f"Kernel not found at {kernel}")
+
+    firmware = finalize_qemu_firmware(config, kernel)
+
+    if firmware == QemuFirmware.linux:
+        kernel = config.output_dir_or_cwd() / config.output_split_kernel
+        if not kernel.exists():
+            die(
+                f"Kernel or UKI not found at {kernel}",
+                hint="Please install a kernel in the image or provide a --qemu-kernel argument to mkosi vmspawn"
+            )
+
+    cmdline: list[PathString] = [
+        "systemd-vmspawn",
+        "--qemu-smp", config.qemu_smp,
+        "--qemu-mem", config.qemu_mem,
+        "--qemu-kvm", config.qemu_kvm.to_tristate(),
+        "--qemu-vsock", config.qemu_vsock.to_tristate(),
+        "--vtpm", config.qemu_swtpm.to_tristate(),
+        "--secure-boot", yes_no(config.secure_boot),
+    ]
+
+    if config.qemu_gui:
+        cmdline += ["--qemu-gui"]
+
+    cmdline += [f"--set-credential={k}:{v}" for k, v in config.credentials.items()]
+
+    with contextlib.ExitStack() as stack:
+        if config.ephemeral:
+            fname = stack.enter_context(copy_ephemeral(config, config.output_dir_or_cwd() / config.output))
+        else:
+            fname = config.output_dir_or_cwd() / config.output
+
+        if config.output_format == OutputFormat.disk and config.runtime_size:
+            apply_runtime_size(fname, config.runtime_size)
+
+        if (
+            kernel and
+            (KernelType.identify(kernel) != KernelType.uki or not config.architecture.supports_smbios(firmware))
+        ):
+            kcl = config.kernel_command_line + config.kernel_command_line_extra
+        else:
+            kcl = config.kernel_command_line_extra
```

Comment:
I think we should just pass `config.kernel_command_line_extra` to vmspawn. Doing it like this doesn't really help since vmspawn will only use SMBIOS to pass extra kernel cmdline arguments anyway.

Reply:
Fixed in 86bd9a91.